lib/Basic/SourceManager.cpp

```diff
@@ -1,39 +1,225 @@
 #include "Basic/SourceManager.hpp"
+#include "Basic/SourceLocation.hpp"
 
 llvm::ErrorOr<glu::FileID> glu::SourceManager::loadFile(llvm::StringRef filePath
 )
 {
     llvm::ErrorOr<std::unique_ptr<llvm::vfs::File>> file
         = _vfs->openFileForRead(filePath);
-
     if (!file) {
         return file.getError();
     }
 
     llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> buffer
         = (*file)->getBuffer(filePath);
-
     if (!buffer) {
         return buffer.getError();
     }
 
     auto fileName = (*file)->getName();
-
     if (!fileName) {
         return fileName.getError();
     }
 
-    uint32_t newOffset = _fileLocEntries.empty()
-        ? 0
-        : _fileLocEntries.back().getOffset() + (*buffer)->getBufferSize();
+    uint32_t fileOffset = _nextOffset;
+    uint32_t fileSize = (*buffer)->getBufferSize();
+    _nextOffset += fileSize;
 
     _fileLocEntries.emplace_back(
-        newOffset, std::move(*buffer), SourceLocation(0), *fileName
+        fileOffset, std::move(*buffer), SourceLocation(fileOffset), *fileName
     );
 
+    if (_fileLocEntries.size() == 1) {
+        _mainFile = FileID(0);
+    }
+
+    return llvm::ErrorOr<glu::FileID>(glu::FileID(_fileLocEntries.size() - 1));
+}
+
+llvm::MemoryBuffer *glu::SourceManager::getBuffer(FileID fileId) const
+{
+    if (fileId._id >= _fileLocEntries.size()) {
+        return nullptr;
+    }
+
+    auto const &entry = _fileLocEntries[fileId._id];
+    std::optional<llvm::MemoryBufferRef> bufferOpt = entry.getBufferIfLoaded();
+
+    if (!bufferOpt) {
+        return nullptr;
+    }
+
+    return entry._buffer.get();
+}
+
+glu::FileID glu::SourceManager::getFileID(uint32_t offset) const
+{
     if (_fileLocEntries.empty()) {
```

Comment:
this if is useless, if it's empty it won't go in the for and will return -1 anyway